pandas/io/excel.py

```diff
@@ -387,7 +388,10 @@ def __init__(self, io, **kwds):
             self.book = io
         elif not isinstance(io, xlrd.Book) and hasattr(io, "read"):
             # N.B. xlrd.Book has a read attribute too
-            if hasattr(io, 'seek'):
+            #
+            # Excluding the case http.client.HTTPResponse.seek() here
+            # because it raises an UnsupportedOperation exception
+            if not isinstance(io, HTTPResponse) and hasattr(io, 'seek'):
```

Comment:
I worry a bit about this hardcoding `HTTPResponse` here.

IIUC, `io` could be anything that implements the `__fspath__` protocol. Not all of those objects will implement `seek`.

What's the harm in cactching `(OSError, ValueError)` here? (I don't really know why we seek in the first place)

Reply:
Back in #19779 we were getting very intermittent failures on ASV tests due to a `TypeError` being thrown. tbh I'm still not entirely clear on *why* it was happening but I believe that intermittently that buffer was being read before this code was being executed. Adding in the conditional seek back to the beginning it prevented the `TypeError`.

So it's certainly a fair question on your end to which I don't have a great answer. try...catch could be an option, but then we'd probably have to seek back to the start in the catch block and maybe still have the conditional (to distinguish from other `TypeError`s) so not sure there is a ton to gain there